updated MIDI input handling and user interface. added features and fixed a couple issues

beau-seidon opened the pull request:
MIDI note and velocity input is now buffered to facilitate legato playing, and polyphony.
Notes are remembered in the order played, limited only by free memory.

Fixes #32 

The following new output modes are added:

Note1, Note2, Note3, Note4:
    V/Oct CV from the 4 most recent notes can be sent to individual outputs.
    When less than 4 notes are played, notes are distributed somewhat evenly.
    Eg., when only 2 notes are held, outputs 1 and 3 play the newer note, outputs 2 and 4 play the older note.

LoNote, HiNote:
    V/Oct corresponding to the lowest and highest held note, pretty simple.

PdlNote:
    V/Oct corresponding to the ["pedal" note](https://en.wikipedia.org/wiki/Pedal_point), or the earliest played note. If this note is released, the next earliest is played.
    Useful as a drone over which other notes can be performed from a 2nd output.

InvNote:
    V/Oct output corresponding to an "inverted" keyboard. High keys play low notes, and low keys play high notes. 
    (127 - MidiNote)

Trig1st:
    Same as Trig, but only triggers on the 1st played note. Useful for interesting 1-shot effects.

TrgAlws:
    Same as Trig, but also triggers when notes are released.

GateInv:
    Same as Gate, but high with no notes held, and low when notes are held.
    Tip: Try setting 2 outputs as Gate and GateInv respectively, then trigger 2 different voices in stereo.

Veloc1, Veloc2, Veloc3, Veloc4:
    Outputs the velocity from the last 4 played notes. 


Applet UI has been updated to allow more characters, and to fix the overlap issue with the right hemisphere activity monitor icon in MIDIOut. MIDIIn now monitors MIDI activity per ADC channel.
Logging has been updated to prevent duplicate note entries on the same MIDI channel.


I think some of the stuff at the bottom of HSIOFrame.h is showing changed because of the whitespace differences.

beau-seidon committed:
note buffer improves playability. midi input log now ignores duplicates

added note buffer to improve playability

stuck velocity bug fixed, and code clean-up

note buffer is now a std::vector for simplicty and memory conservation.
might have broken note-off logging

fixed logging bug from last commit, also now prevents duplicate log entries.
moved some stuff around for readability

beau-seidon committed:
fixed midi monitor icon overlap on right HS

beau-seidon committed:
added assignable polyphonic note output

hMIDIIn polyphonic cv output working! saving broken...

updated polyphonic note priority. saving still broken

fixed polyphonic note-off bug

saving is fixed! thanks djphazer

beau-seidon committed:
added additional output modes, and updated ui

added more note out functions, and improved code readability

adjusted ClockSetup SLIDEOUT_TIME

more gate and trig features. updated UI for more characters

beau-seidon commented:
things i haven't tested thoroughly:
- NoteStackClear(), triggered by usbMIDI.Stop
- midi input passed to other applets on the back-end

beau-seidon commented:
i also intend to attempt the per-channel activity monitor icon in my next commit, so don't merge just yet
EDIT: done

beau-seidon commented:
ALSO, one thing I still don't like is that the new modes I put in are separate in the list from the others of the same type. 
By that I mean the new trigger and gate modes are at the end of the list instead of up near the original Trig and Gate modes, etc. 
But I'm worried that if I start shuffling things around in the MIDIFunctions enum that I'll break things that worked before i started rummaging around.

EDIT: rearranged and tested, all good. i like it better how i have it now, but i'm not the only one who uses the thing. also i didn't rename HEM_MIDI_NOTE_OUT in the enum because it is used by other things i didn't modify.

djphazer commented:
Haha no worries, we'll let this one cook for a while. I was going to say, we don't need the old "Note#" output mode if it's identical to the new "Poly1" mode... I'd probably just rename those to "Note1", "Note2", etc.

If you reorder the enum, I think you just have to reorder the string labels as well... everything else should make safe assumptions, using switch statements, etc.

beau-seidon committed:
midi activity monitor is now per adc channel. added poly velocity out

added midi input monitoring per ADC channel, with midi channel filtering

updated labels and order of midi functions

added "polyphonic" velocity output
also fixed note-off poly3 behavior and activity icon overlap with log view

beau-seidon commented:
updated to include per channel monitoring, and fixed the mode list. let me know what you think of the updated applet!
~i think i'm done with it for now, unless there are bugs~

EDIT: nope, i think ill add the poly velocity afterall.

beau-seidon commented:
ok, added polyphonic velocity. its fun. 
also fixed a graphical boo boo, and improved Note3-off behavior.

i'm done. do whatever you want with it. :)

djphazer commented:
Excellent! We'll give this some testing to put it through its paces... I briefly tried out Note1 and Note2 with two different voices, and it seems solid!

For the sake of tidiness, I'd prefer to have this rebased against the main branch (drop the merge commit) and maybe squashed into fewer commits if possible... something like `git rebase -i phazerville` to rewrite the history, and then force-push to your feature branch after cleanup.

beau-seidon commented:
I'll give it a shot! i always seem to screw up a rebase tho. i suck at git. the only reason i can do any of it properly is the Source Control extension for VSCode. we'll see what happens. 

do you want just a few major ones, without all the piddly "fixed a bug from last commit" stuff, or just squash the whole damn thing?

djphazer commented:
Oh nooo, we're gonna have to get you acquainted with git in a terminal...
You could squash the whole thing into one, with notes in the commit message. But yeah, we don't need all the little fixups, just the high-level atomic feature!

beau-seidon commented:
im trying this one in a terminal. gotta saw off the training wheels.

beau-seidon commented:
`git rebase -i` opened in vim and now im mad at you.

beau-seidon committed:
MIDIIn now supports polyphonic key aftertouch

beau-seidon committed:
MIDIIn now automatically handles sustain pedal

beau-seidon commented:
are you opposed to me closing this PR? all of it (and more) has already been cherry-picked into 1.9. 
tryna purge all my extra branches. :)

djphazer commented:
yeah, no worries! I've stolen it all behind the scenes. GitHub PRs are just formalities. ;)
